pkg/cloudprovider/providers/gce/gce_kms.go

```diff
@@ -0,0 +1,189 @@
+}
+
+func init() {
+	cloudprovider.RegisterKMSService(
+		ServiceName,
+		func(cloud cloudprovider.Interface, config map[string]interface{}) (cloudprovider.KMSService, error) {
+			return newGoogleKMSService(cloud, config)
+		})
+}
+
+// gkmsService provides Encrypt and Decrypt methods which allow cryptographic operations
+// using Google Cloud KMS service.
+type gkmsService struct {
+	parentName      string
+	cloudkmsService *cloudkms.Service
+}
+
+// newGoogleKMSService creates a Google KMS connection and returns a kms.Service instance which can encrypt and decrypt data.}
+func newGoogleKMSService(cloud cloudprovider.Interface, rawConfig map[string]interface{}) (cloudprovider.KMSService, error) {
+	var cloudkmsService *cloudkms.Service
+	var cloudProjectID string
+
+	// This check is false if cloud is nil, or is not an instance of gce.GCECloud.
+	if gke, ok := cloud.(*GCECloud); ok {
+		// Hosting on GCE/GKE with Google KMS encryption provider
+		cloudkmsService = gke.GetKMSService()
+
+		// Project ID is assumed to be the user's project unless there
+		// is an override in the configuration file. If there is an override,
+		// it will be taken into account by the Google KMS service constructor,
+		// after reading the configuration file.
+		cloudProjectID = gke.ProjectID()
+	} else {
+		// When running outside GCE/GKE and connecting to KMS, GOOGLE_APPLICATION_CREDENTIALS
+		// environment variable is required. This describes how that can be done:
+		// https://developers.google.com/identity/protocols/application-default-credentials
+		ctx := context.Background()
+		client, err := google.DefaultClient(ctx, cloudkms.CloudPlatformScope)
+		if err != nil {
+			return nil, err
+		}
+		cloudkmsService, err = cloudkms.New(client)
+		if err != nil {
+			return nil, err
+		}
+		cloudProjectID = ""
+	}
+
+	var config KMSConfig
+	err := mapstructure.Decode(rawConfig, &config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set defaults for projectID, location and keyRing.
+	if len(config.ProjectID) == 0 {
+		config.ProjectID = cloudProjectID
+	}
+	if len(config.Location) == 0 {
+		config.Location = "global"
+	}
+	if len(config.KeyRing) == 0 {
+		config.KeyRing = defaultGKMSKeyRing
+	}
+
+	if len(config.ProjectID) == 0 {
+		return nil, fmt.Errorf("missing projectID in encryption provider configuration for gkms provider")
+	}
+	if len(config.CryptoKey) == 0 {
+		return nil, fmt.Errorf("missing cryptoKey in encryption provider configuration for gkms provider")
+	}
+
+	parentName := fmt.Sprintf("projects/%s/locations/%s", config.ProjectID, config.Location)
+
+	// Create the keyRing if it does not exist yet
+	_, err = cloudkmsService.Projects.Locations.KeyRings.Create(parentName,
+		&cloudkms.KeyRing{}).KeyRingId(config.KeyRing).Do()
+	if err != nil {
+		apiError, ok := err.(*googleapi.Error)
+		// If it was a 409, that means the keyring existed.
+		// If it was a 403, we do not have permission to create the keyring, the user must do it.
```

Comment:
If we do not have permission to create the keyring and the user must do it, don't they have to do it before we can proceed?